File: Sources/AndroidBackend/Kotlin/ViewOnHoverListener.kt
package dev.swiftcrossui.androidbackend

import android.view.View
import android.view.MotionEvent

class ViewOnHoverListener(
    private val enterAction: SwiftAction,
    private val leaveAction: SwiftAction
) : View.OnHoverListener {
    override fun onHover(view: View, event: MotionEvent): Boolean {
        if (event.getAction() == MotionEvent.ACTION_HOVER_ENTER) {
            enterAction.call()
            return true
        }
        
        if (event.getAction() == MotionEvent.ACTION_HOVER_EXIT) {
            leaveAction.call()
            return true
        }

        return false
    }
}

File: Sources/AndroidBackend/ViewOnHoverListener.swift
import AndroidKit
import SwiftJava

@JavaClass(
    "dev.swiftcrossui.androidbackend.ViewOnHoverListener",
    extends: AndroidView.View.OnHoverListener.self
)
class ViewOnHoverListener: JavaObject {
    @JavaMethod
    @_nonoverride convenience init(
        enterAction: SwiftAction?,
        leaveAction: SwiftAction?,
        environment: JNIEnvironment? = nil
    )
}

extension ViewOnHoverListener {
    convenience init(
        enterAction: @escaping () -> (),
        leaveAction: @escaping () -> (),
        environment: JNIEnvironment? = nil
    ) {
        let enterObject = SwiftAction(
            environment: environment,
            action: enterAction
        )
        let leaveObject = SwiftAction(
            environment: environment,
            action: leaveAction
        )
        self.init(
            enterAction: enterObject,
            leaveAction: leaveObject,
            environment: environment
        )
    }
}
